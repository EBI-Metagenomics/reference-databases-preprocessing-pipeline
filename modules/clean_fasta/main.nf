
process CLEAN_FASTA {

    label 'light'
    container '/hps/nobackup/rdf/metagenomics/service-team/singularity-cache/quay.io_biocontainers_seqtk:1.3.sif'
    publishDir "${params.outdir}/${label}/${version}/", mode: 'copy'

    input:
    path fasta
    path tax
    val version
    val label

    output:
    path("${label}.fasta"), includeInputs: true, emit: cleaned_fasta

    script:
    """
    case $label in

        UNITE)
            sed 's/|.*//g' $fasta > ${fasta}.clean
            grep -v '^#' $tax | cut -f1 > ${label}.idlst
            seqtk subseq ${fasta}.clean ${label}.idlst > ${label}_subseq.fasta
            sed 's/NN*/N/g' ${label}_subseq.fasta > ${label}.fasta
        ;;
        
        ITSone)
            sed "s/|ITS1 located by ENA annotation,.*//g" $fasta > ${fasta}_temp.clean
            sed 's/ /_/g' ${fasta}_temp.clean > ${fasta}.clean
            grep -v '^#' $tax | cut -f1 > ${label}.idlst
            seqtk subseq ${fasta}.clean ${label}.idlst > ${label}_temp.fasta
            sed 's/ /_/g' ${label}_temp.fasta > ${label}.fasta

        ;;

        SILVA-SSU | SILVA-LSU)
            grep -v '^#' $tax | cut -f1 > ${label}.idlst
            seqtk subseq $fasta ${label}.idlst > ${label}.fasta
        ;;

        PR2)
            echo "No cleaning necessary."
        ;;

        *)
            echo "Incorrect reference db label"
            exit 1
        ;;
        
    esac
    """

}