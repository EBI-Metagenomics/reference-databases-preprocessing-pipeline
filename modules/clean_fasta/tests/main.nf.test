nextflow_process {

    name "Test Process CLEAN_FASTA"
    script "modules/clean_fasta/main.nf"
    process "CLEAN_FASTA"
    tag "modules"
    tag "clean_fasta"

    test("Test clean_fasta process for SILVA") {
        tag "clean_fasta_silva"

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = file('${projectDir}/modules/clean_fasta/tests/data/SILVA/test.fasta', checkIfExists: true)
                input[1] = file('${projectDir}/modules/clean_fasta/tests/data/SILVA/test-tax.txt', checkIfExists: true)
                input[2] = params.silva_version
                input[3] = params.silva_ssu_label
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.cleaned_fasta).match("silva_cleaned_fasta") }
            )
        }

    }

    test("Test clean_fasta process for UNITE") {
        tag "clean_fasta_unite"

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = file('${projectDir}/modules/clean_fasta/tests/data/UNITE/test.fasta', checkIfExists: true)
                input[1] = file('${projectDir}/modules/clean_fasta/tests/data/UNITE/test-tax.txt', checkIfExists: true)
                input[2] = params.unite_version
                input[3] = params.unite_label
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.cleaned_fasta).match("unite_cleaned_fasta") }
            )
        }

    }

    test("Test clean_fasta process for ITSONEdb") {
        tag "clean_fasta_itsonedb"

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = file('${projectDir}/modules/clean_fasta/tests/data/ITSONEdb/test.fasta', checkIfExists: true)
                input[1] = file('${projectDir}/modules/clean_fasta/tests/data/ITSONEdb/test-tax.txt', checkIfExists: true)
                input[2] = params.itsonedb_version
                input[3] = params.itsonedb_label
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.cleaned_fasta).match("itsonedb_cleaned_fasta") }
            )
        }

    }

}
