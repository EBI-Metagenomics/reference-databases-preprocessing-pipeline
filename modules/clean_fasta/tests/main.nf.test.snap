{
    "unite_cleaned_fasta": {
        "content": [
            [
                "UNITE.fasta:md5,e948ac7dd52ed9374b99fe0071a404bc"
            ]
        ],
        "timestamp": "2023-12-12T12:17:03.447478493"
    },
    "itsonedb_cleaned_fasta": {
        "content": [
            [
                "ITSone.fasta:md5,3674a76b631981ae2cfc7ef456951db7"
            ]
        ],
        "timestamp": "2023-12-12T12:17:13.221527134"
    },
    "silva_cleaned_fasta": {
        "content": [
            [
                "SILVA-SSU.fasta:md5,68c60c9c9b78c0244f05e78fac9926a6"
            ]
        ],
        "timestamp": "2023-12-12T12:16:23.566759315"
    }
}